### DISCUSSION TITLE
Array, List, Collection, Set, ReadOnlyList - what? A comprehensive and exhaustive list of collection-like types

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#36] [General]
# Array, List, Collection, Set, ReadOnlyList - what? A comprehensive and exhaustive list of collection-like types

.NET knows a big list of collection-like types like: IEnumerable, IQueryable, IList, ICollection, Array, ISet, ImmutableArray, ReadOnlyCollection, ReadOnlyList, and many more.
This blog post will give you an exhaustive list of types in .NET and when to use what.

https://steven-giesel.com/blogPost/f11aadf5-5a55-4742-b046-28dddd0b3ebf

**Comment by icodev-space** [comment]
A great reference material on collections, concise and well put together, thanks.

**Comment by ahmedanwar100** [comment]
Very unclear about these lines in LinkedList section:
Also if a List is very big (>85kb) it goes onto the large object heap - LinkedList wouldn't
They can help with fragmentation

**Reply by linkdotnet** [reply]
Thanks for the input - I added a link to another blog post of mine giving some more context.
Here in short: A `List` is basically one array - if that grows over 85kb in size the Garbage Collector puts it onto the Large object heap. A LinkedList consists of nodes - all in all, they can have more than 85kb in size but a single node is most probably very small. So your `LinkedList` will not be part of the Large object heap. Why LOC exists and stuff is explained in the linked blog post)

**Comment by ahmedanwar100** [comment]
And also unclear about IImmutableList and FrozenSet difference and usage

**Reply by linkdotnet** [reply]
I can make this also a bit more clear. Basically an immutable list and a frozen set are "almost" the same.
Immutable list allows you to add an entry to your current instance like this:
```csharp
var myNewImmutableList.= immutableList.Add(2);
```
So you get a new instance and you did not modify your original object. Now those two objects share every entry except the newly added one. Therefore almost all immutable implementations have some complex mechanism to share the memory - reducing footprint and increasing runtime. A Frozen Set does not have such capabilities - it is basically like an immutable set without the add, remove functions, making it way easier to implement and doing certain stuff.

So if you are absolutely sure that your collection will **always** stay the same, the new type would be your goto option.

**Comment by diegodrf** [comment]
Thank you for this article. 
You are a fantastic writer/teacher. 
I know that I will come back to this topic many times to find the best type for my code.

**Reply by linkdotnet** [reply]
Hey @diegodrf. Thank you very much for the supportive words. That really means a lot to me!

**Comment by diegodrf** [comment]
Hi @linkdotnet,

To keep this amazing post up to date, in C# 13 there's a new feature about [params](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-13#params-collections). 

On the topic **Array** -> **When to use?** -> **3** 
It may not be a decisive criterion for the use of Array depending on the language version.

**Reply by linkdotnet** [reply]
That's a good point. Once a I have a bit more time I'll update this. Thanks @diegodrf
